pkg/agent/util/net_windows.go

```diff
@@ -655,673 +56,14 @@ func ListenLocalSocket(address string) (net.Listener, error) {
-			ifi.Flags |= net.FlagBroadcast | net.FlagMulticast
-		case windows.IF_TYPE_PPP, windows.IF_TYPE_TUNNEL:
-			ifi.Flags |= net.FlagPointToPoint | net.FlagMulticast
-		case windows.IF_TYPE_SOFTWARE_LOOPBACK:
-			ifi.Flags |= net.FlagLoopback | net.FlagMulticast
-		case windows.IF_TYPE_ATM:
-			ifi.Flags |= net.FlagBroadcast | net.FlagPointToPoint | net.FlagMulticast // assume all services available; LANE, point-to-point and point-to-multipoint
-		}
-		if aa.Mtu == 0xffffffff {
-			ifi.MTU = -1
-		} else {
-			ifi.MTU = int(aa.Mtu)
-		}
-		if aa.PhysicalAddressLength > 0 {
-			ifi.HardwareAddr = make(net.HardwareAddr, aa.PhysicalAddressLength)
-			copy(ifi.HardwareAddr, aa.PhysicalAddress[:])
-		}
-		adapter := adapter{
-			Interface:     ifi,
-			compartmentID: aa.CompartmentId,
-			flags:         aa.Flags,
-		}
-		adapters = append(adapters, adapter)
-	}
-	return adapters, nil
+func SetAdapterMACAddress(adapterName string, macConfig *net.HardwareAddr) error {
+	return winnet.SetAdapterMACAddress(adapterName, macConfig)
 }
 
-// GAA_FLAG_INCLUDE_ALL_COMPARTMENTS is used in windows.GetAdapterAddresses parameter
-// flags to return addresses in all routing compartments.
-const GAA_FLAG_INCLUDE_ALL_COMPARTMENTS = 0x00000200
-
-// GAA_FLAG_INCLUDE_ALL_INTERFACES is used in windows.GetAdapterAddresses parameter
-// flags to return addresses for all NDIS interfaces.
-const GAA_FLAG_INCLUDE_ALL_INTERFACES = 0x00000100
-
-// adapterAddresses returns a list of IpAdapterAddresses structures. The structure
-// contains an IP adapter and flattened multiple IP addresses including unicast, anycast
-// and multicast addresses.
-// This function is copied from go/src/net/interface_windows.go, with a change that flag
-// GAA_FLAG_INCLUDE_ALL_COMPARTMENTS is introduced to query interfaces in all compartments,
-// and GAA_FLAG_INCLUDE_ALL_INTERFACES is introduced to query all NDIS interfaces even they
-// are not configured with any IP addresses, e.g., uplink.
-func adapterAddresses() ([]*windows.IpAdapterAddresses, error) {
-	flags := uint32(windows.GAA_FLAG_INCLUDE_PREFIX | GAA_FLAG_INCLUDE_ALL_COMPARTMENTS | GAA_FLAG_INCLUDE_ALL_INTERFACES)
-	var b []byte
-	l := uint32(15000) // recommended initial size
-	for {
-		b = make([]byte, l)
-		err := getAdaptersAddresses(syscall.AF_UNSPEC, flags, 0, (*windows.IpAdapterAddresses)(unsafe.Pointer(&b[0])), &l)
-		if err == nil {
-			if l == 0 {
-				return nil, nil
-			}
-			break
-		}
-		if err.(syscall.Errno) != syscall.ERROR_BUFFER_OVERFLOW {
-			return nil, os.NewSyscallError("getadaptersaddresses", err)
-		}
-		if l <= uint32(len(b)) {
-			return nil, os.NewSyscallError("getadaptersaddresses", err)
-		}
-	}
-	var aas []*windows.IpAdapterAddresses
-	for aa := (*windows.IpAdapterAddresses)(unsafe.Pointer(&b[0])); aa != nil; aa = aa.Next {
-		aas = append(aas, aa)
-	}
-	return aas, nil
+func SetLinkUp(name string) (net.HardwareAddr, int, error) {
+	return winnet.SetLinkUp(name)
```

Comment:
For these functions, I am still not sure I understand the value of moving the actual implementation to the `winnet` package. I could understand if `winnet.SetLinkUp` was part of `winnet.Interface` but that's not the case.

Overall, the way we split functions is a bit confusing to me, but maybe a PR description would clarify that. Wouldn't it be more natural to have all the functions is in the interface, to mirror what we do with `/pkg/agent/util/netlink/`.

Reply:
I have made another change but not finished it. Not sure if the change is the expectation you said.

Reply:
I moved all the functions that use Powershell or Windows syscall to `/pkg/agent/util/winnet`. The `high-level` functions that call the functions moved to `/pkg/agent/util/winnet` are reserved. Besides, in order to not change the files that use the functions moved to `/pkg/agent/util/winnet`, I added some functions, like the following one, that only call the functions with the same name in `/pkg/agent/util/winnet`, though the corresponding functions can be called directly by importing `/pkg/agent/util/winnet`.

```go
func VMSwitchExists() (bool, error) {
	return winnetUtil.VMSwitchExists(LocalVMSwitch)
}
```

Reply:
Yes, this seems to be more like what I had in mind, with all functions added to `winnet.Interface`.
We should get a review from @tnqn and @wenyingd as well to see what they think of the latest version.